streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```diff
@@ -195,6 +195,16 @@
      */
     public static final String ADMIN_CLIENT_PREFIX = "admin.";
 
+    /**
+     * Config value for parameter (@link #TOPOLOGY_OPTIMIZATION "topology.optimization" for disabling topology optimization
+     */
+    public static final String NO_OPTIMIZATION = "none";
+
+    /**
+     * Config value for parameter (@link #TOPOLOGY_OPTIMIZATION "topology.optimization" for enabling topology optimization
+     */
+    public static final String OPTIMIZE = "all";
```

Comment:
I know we've been around and around about this, and at this point, I can't remember if this was already discussed and discarded, but...

I think it may make sense to just add "none" or" 2.0" as the possible values. 

We still aren't sure if upgrading optimizations will always be safe, so explicitly stating the optimization value you want will protect you against accidental upgrades when 2.1+ come out.

Also, the explicit optimization level will protect against accidental *downgrades*, which could also be unsafe. For example, if I deploy v2.1 with "all", it'll set me up with 2.1-level optimizations. If I decide I don't like it and go down to 2.0, still with "all", Streams won't know any different, so it'll construct my topology according to 2.0 rules, potentially leading to data loss.

On the other hand, if I had deployed it with "2.1" and then downgraded to 2.0, Streams could refuse to start, since it doesn't understand that config value.

I think it would be really nice, in the future, to investigate ways to restructure the topology dynamically without losing data, but until we get there, it's better to err on the side of explicitness and safety.

Reply:
I agree that adding for "2.0" would make sense for upgrading / downgrading; in fact, I'm working on some optimizations for allowing users to decide whether to turn on / off source KTables based on the config, which would require a "2.0" value.

On the other hand, I think "all" still have its value, when 1) I am in development mode and do not care about compatibility, or 2) I have used other mechanisms to makes sure the persistent storages like Kafka topics would not change (this is future work).

So I'm +1 to merge this PR as is, and work on adding another value of "2.0" in a separate PR.

Reply:
I think "all" is fine, as "all" always means "latest version" -- if we release 2.1, we would introduce "2.0" and "all" as become a surrogate for "2.1" -- I think it's clean this way, ie, we only have a version string for older releases.

Reply:
I'm +1 with "all" for the latest release and a version string for older releases.